File: client/common/protocol.go
package common

import (
	"encoding/binary"
	"fmt"
	"net"
	"strconv"
	"strings"
)

// SerializeMessage converts a message to custom protocol format
func SerializeMessage(message interface{}) ([]byte, error) {
	var data []byte

	switch msg := message.(type) {
	case *BetMessage:
		data = append(data, byte(MessageTypeBet))
		content := fmt.Sprintf("%s|%s|%s|%s|%d", msg.Nombre, msg.Apellido, msg.Documento, msg.Nacimiento, msg.Numero)
		data = append(data, []byte(content)...)

	case *BatchMessage:
		data = append(data, byte(MessageTypeBatch))
		content := fmt.Sprintf("%d|%d|%d", msg.Agency, boolToInt(msg.EOF), len(msg.Bets))
		for _, bet := range msg.Bets {
			content += fmt.Sprintf("|%s|%s|%s|%s|%d", bet.Nombre, bet.Apellido, bet.Documento, bet.Nacimiento, bet.Numero)
		}
		data = append(data, []byte(content)...)

	case *GetWinnersMessage:
		data = append(data, byte(MessageTypeGetWinners))
		content := fmt.Sprintf("%d", msg.Agency)
		data = append(data, []byte(content)...)

	default:
		return nil, fmt.Errorf("unsupported message type")
	}

	return data, nil
}

// DeserializeMessage parses custom protocol data into a message
func DeserializeMessage(data []byte) (interface{}, error) {
	if len(data) < 1 {
		return nil, fmt.Errorf("empty message")
	}

	msgType := MessageType(data[0])
	content := string(data[1:])

	switch msgType {
	case MessageTypeResponse:
		return parseResponseMessage(content)
	default:
		return nil, fmt.Errorf("unsupported message type: %d", msgType)
	}
}

// parseResponseMessage parses response message from pipe-delimited content
func parseResponseMessage(content string) (*ResponseMessage, error) {
	parts := strings.Split(content, "|")
	if len(parts) < 3 {
		return nil, fmt.Errorf("invalid response format")
	}

	success := parts[0] == "1"
	errorMsg := parts[1]

	winnerCount, err := strconv.Atoi(parts[2])
	if err != nil {
		return nil, fmt.Errorf("invalid winner count: %w", err)
	}

	var winners []string
	for i := 3; i < 3+winnerCount && i < len(parts); i++ {
		if parts[i] != "" {
			winners = append(winners, parts[i])
		}
	}

	return &ResponseMessage{
		Type:    MessageTypeResponse,
		Success: success,
		Error:   errorMsg,
		Winners: winners,
	}, nil
}

// SendMessage serializes message and sends with length prefix
func SendMessage(conn net.Conn, message interface{}) error {
	data, err := SerializeMessage(message)
	if err != nil {
		return fmt.Errorf("error serializing message: %w", err)
	}

	// Send length prefix
	length := uint32(len(data))
	lengthBytes := make([]byte, LENGTH_PREFIX_BYTES)
	binary.BigEndian.PutUint32(lengthBytes, length)

	if _, err := conn.Write(lengthBytes); err != nil {
		return fmt.Errorf("error sending length: %w", err)
	}

	// Send message data
	if _, err := conn.Write(data); err != nil {
		return fmt.Errorf("error sending data: %w", err)
	}

	return nil
}

// RecvMessage reads length-prefixed message and deserializes
func RecvMessage(conn net.Conn, v interface{}) error {
	// Read length prefix
	lengthBytes := make([]byte, LENGTH_PREFIX_BYTES)
	if _, err := readExact(conn, lengthBytes); err != nil {
		return fmt.Errorf("error reading length: %w", err)
	}

	length := binary.BigEndian.Uint32(lengthBytes)

	// Read message data
	data := make([]byte, length)
	if _, err := readExact(conn, data); err != nil {
		return fmt.Errorf("error reading data: %w", err)
	}

	// Deserialize message
	message, err := DeserializeMessage(data)
	if err != nil {
		return fmt.Errorf("error deserializing message: %w", err)
	}

	// Copy to destination
	switch dst := v.(type) {
	case *ResponseMessage:
		if src, ok := message.(*ResponseMessage); ok {
			*dst = *src
		} else {
			return fmt.Errorf("type mismatch: expected ResponseMessage")
		}
	default:
		return fmt.Errorf("unsupported destination type")
	}

	return nil
}

// readExact reads exactly len(buf) bytes from conn
func readExact(conn net.Conn, buf []byte) (int, error) {
	totalRead := 0
	for totalRead < len(buf) {
		n, err := conn.Read(buf[totalRead:])
		if err != nil {
			return totalRead, err
		}
		totalRead += n
	}
	return totalRead, nil
}

File: server/protocol/protocol.py
from .messages import (
    BetMessage,
    BatchMessage,
    GetWinnersMessage,
    MESSAGE_TYPE_BET,
    MESSAGE_TYPE_BATCH,
    MESSAGE_TYPE_GET_WINNERS,
    MESSAGE_TYPE_RESPONSE,
)

BUFFER_SIZE = 4096


def read_packet_from(client_socket):
    """Read length-prefixed packet and parse message"""
    # Read length prefix (4 bytes)
    length_data = _read_exact(client_socket, 4)
    length = int.from_bytes(length_data, byteorder="big")
    # Read message data
    data = _read_exact(client_socket, length)

    # Parse based on message type
    if len(data) < 1:
        raise ValueError("Empty message")

    msg_type = data[0]

    if msg_type == MESSAGE_TYPE_BET:
        return BetMessage.from_data(data)
    elif msg_type == MESSAGE_TYPE_BATCH:
        return BatchMessage.from_data(data)
    elif msg_type == MESSAGE_TYPE_GET_WINNERS:
        return GetWinnersMessage.from_data(data)
    else:
        raise ValueError(f"Unknown message type: {msg_type}")


def send_response(client_socket, success, error=None, winners=None):
    """Send response message using custom protocol"""
    # Build response data
    data = bytearray()
    data.append(MESSAGE_TYPE_RESPONSE)

    # Add success flag
    data.extend(("1" if success else "0").encode("utf-8"))
    data.extend(b"|")

    # Add error if present
    if error:
        data.extend(error.encode("utf-8"))
    data.extend(b"|")

    # Add winners if present
    if winners:
        data.extend(str(len(winners)).encode("utf-8"))
        data.extend(b"|")
        for winner in winners:
            data.extend(winner.encode("utf-8"))
            data.extend(b"|")
    else:
        data.extend(b"0|")

    # Send length-prefixed message
    length = len(data)
    length_bytes = length.to_bytes(4, byteorder="big")

    _send_exact(client_socket, length_bytes)
    _send_exact(client_socket, data)


def _send_exact(sock, data):
    """Send exactly all bytes in data (prevents short write)"""
    total_sent = 0
    while total_sent < len(data):
        sent = sock.send(data[total_sent:])
        if sent == 0:
            raise RuntimeError("Socket connection broken")
        total_sent += sent


def _read_exact(sock, n):
    """Read exactly n bytes from socket"""
    chunks = []
    bytes_read = 0
    while bytes_read < n:
        chunk = sock.recv(n - bytes_read)
        if not chunk:  # Connection closed
            raise RuntimeError("Socket connection broken")
        chunks.append(chunk)
        bytes_read += len(chunk)
    return b"".join(chunks)
